modules/subchannel/src/problems/SubChannel1PhaseProblem.C

```diff
@@ -1906,215 +1953,112 @@ SubChannel1PhaseProblem::computeAddedHeatDuct(unsigned int i_ch, unsigned int iz
+  // ---------- helper functions -----------------------------
+  auto V = [&](const std::string & s)
   {
-    mat_array[Q * field_num + 3] = NULL;
-  }
-  LibmeshPetscCall(VecDuplicate(_mc_axial_convection_rhs, &vec_array[field_num]));
-  LibmeshPetscCall(VecCopy(_mc_axial_convection_rhs, vec_array[field_num]));
-  if (!_axial_mass_flow_tight_coupling)
+    if (_verbose_subchannel)
+      _console << s << std::endl;
+  };
+
+  auto DupMatAssembled = [&](Mat src, Mat * dst)
   {
-    Vec sumWij_loc;
-    LibmeshPetscCall(VecDuplicate(_mc_axial_convection_rhs, &sumWij_loc));
-    LibmeshPetscCall(VecSet(sumWij_loc, 0.0));
-    for (unsigned int iz = first_node; iz < last_node + 1; iz++)
+    if (src)
     {
-      auto iz_ind = iz - first_node;
-      for (unsigned int i_ch = 0; i_ch < _n_channels; i_ch++)
-      {
-        auto * node_out = _subchannel_mesh.getChannelNode(i_ch, iz);
-
-        PetscScalar value_vec_2 = -1.0 * (*_SumWij_soln)(node_out);
-        PetscInt row_vec_2 = i_ch + _n_channels * iz_ind;
-        LibmeshPetscCall(VecSetValues(sumWij_loc, 1, &row_vec_2, &value_vec_2, ADD_VALUES));
-      }
+      LibmeshPetscCall(MatDuplicate(src, MAT_COPY_VALUES, dst));
+      LibmeshPetscCall(MatAssemblyBegin(*dst, MAT_FINAL_ASSEMBLY));
+      LibmeshPetscCall(MatAssemblyEnd(*dst, MAT_FINAL_ASSEMBLY));
     }
-    LibmeshPetscCall(VecAXPY(vec_array[field_num], 1.0, sumWij_loc));
-    LibmeshPetscCall(VecDestroy(&sumWij_loc));
-  }
+    else
+    {
+      *dst = NULL;
+    }
+  };
 
-  if (_verbose_subchannel)
-    _console << "Mass ok." << std::endl;
-  // Axial momentum conservation
-  field_num = 1;
-  if (_pressure_axial_momentum_tight_coupling)
-  {
-    LibmeshPetscCall(
-        MatDuplicate(_amc_sys_mdot_mat, MAT_COPY_VALUES, &mat_array[Q * field_num + 0]));
-    LibmeshPetscCall(MatAssemblyBegin(mat_array[Q * field_num + 0], MAT_FINAL_ASSEMBLY));
-    LibmeshPetscCall(MatAssemblyEnd(mat_array[Q * field_num + 0], MAT_FINAL_ASSEMBLY));
-  }
-  else
+  auto DupVecCopy = [&](Vec src, Vec * dst)
   {
-    mat_array[Q * field_num + 0] = NULL;
-  }
-  LibmeshPetscCall(
-      MatDuplicate(_amc_pressure_force_mat, MAT_COPY_VALUES, &mat_array[Q * field_num + 1]));
-  LibmeshPetscCall(MatAssemblyBegin(mat_array[Q * field_num + 1], MAT_FINAL_ASSEMBLY));
-  LibmeshPetscCall(MatAssemblyEnd(mat_array[Q * field_num + 1], MAT_FINAL_ASSEMBLY));
-  mat_array[Q * field_num + 2] = NULL;
-  if (_monolithic_thermal_bool)
-  {
-    mat_array[Q * field_num + 3] = NULL;
-  }
-  LibmeshPetscCall(VecDuplicate(_amc_pressure_force_rhs, &vec_array[field_num]));
-  LibmeshPetscCall(VecCopy(_amc_pressure_force_rhs, vec_array[field_num]));
-  if (_pressure_axial_momentum_tight_coupling)
-  {
-    LibmeshPetscCall(VecAXPY(vec_array[field_num], 1.0, _amc_sys_mdot_rhs));
-  }
-  else
+    LibmeshPetscCall(VecDuplicate(src, dst));
+    LibmeshPetscCall(VecCopy(src, *dst));
+  };
+
+  // per-field relax+blend helper with explicit return type.
+  auto RelaxField =
```

Comment:
name is bad. this is also relaxing the matrix diagonal not just the field

Reply:
```suggestion
  auto RelaxEquation =
```
same as the other I guess